File: android/src/main/java/com/delicity/thermalprinter/discovery/DiscoveryManager.kt
package com.delicity.thermalprinter.discovery

import android.bluetooth.BluetoothAdapter
import android.content.Context
import com.delicity.thermalprinter.adapters.PrinterAdapter
import com.delicity.thermalprinter.model.DiscoveredPrinter
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withTimeoutOrNull
import java.util.Collections

/**
 * Orchestre la découverte AGRÉGÉE multi-sources en parallèle puis fusionne.
 *
 * Sources lancées (selon options + disponibilité plateforme) :
 *   - SDK Epson / Star / Brother / Zebra (via PrinterAdapter.discover)
 *   - TcpScanner (réseau 9100)
 *   - BluetoothClassicScanner (Android)
 *   - BleScanner (optionnel)
 *   - UsbAdapter.discover (optionnel)
 *
 * Chaque source pousse ses résultats dans un buffer thread-safe ; `emitPartial`
 * est invoqué au fil de l'eau (event printerFound). À la fin : fusion +
 * dédoublonnage + arbitrage d'adapter via AdapterPriority.
 */
class DiscoveryManager(
    private val context: Context,
    private val btAdapter: BluetoothAdapter?,
    private val adapters: List<PrinterAdapter>,
) {

    data class Options(
        val sources: Set<String>?, // null = toutes
        val timeoutMs: Long = 8000,
        val includePaired: Boolean = true,
        val networkCidr: String? = null,
        val tcpPorts: List<Int> = listOf(9100),
    )

    suspend fun discover(
        options: Options,
        emitPartial: (DiscoveredPrinter) -> Unit,
    ): Pair<List<DiscoveredPrinter>, List<String>> = coroutineScope {
        val buffer = Collections.synchronizedList(mutableListOf<DiscoveredPrinter>())
        val failed = Collections.synchronizedList(mutableListOf<String>())

        // Collecte thread-safe + émission partielle (callback synchrone).
        val collect: (DiscoveredPrinter) -> Unit = { p ->
            buffer.add(p)
            runCatching { emitPartial(p) }
        }

        fun enabled(src: String) = options.sources == null || options.sources.contains(src)

        val jobs = mutableListOf<kotlinx.coroutines.Deferred<Unit>>()

        // --- Sources SDK fabricants (via adapters) ---
        for (adapter in adapters) {
            val discoverySource = when (adapter.id.value) {
                "epson" -> "epson"
                "star" -> "star"
                "brother" -> "brother"
                "zebra" -> "zebra"
                else -> null
            } ?: continue
            if (enabled(discoverySource) && adapter.isAvailable()) {
                jobs += async {
                    runCatching {
                        withTimeoutOrNull(options.timeoutMs + 1000) {
                            adapter.discover(options.timeoutMs, collect)
                        }
                    }.onFailure { failed.add(discoverySource) }
                    Unit
                }
            }
        }

        // --- Source TCP réseau ---
        if (enabled("tcp")) {
            jobs += async {
                runCatching {
                    TcpScanner(context).scan(options.timeoutMs, options.tcpPorts, options.networkCidr, collect)
                }.onFailure { failed.add("tcp") }
                Unit
            }
        }

        // --- Source Bluetooth classique (Android) ---
        if (enabled("bluetooth") && btAdapter != null) {
            jobs += async {
                runCatching {
                    BluetoothClassicScanner(context, btAdapter)
                        .scan(options.timeoutMs, options.includePaired, collect)
                }.onFailure { failed.add("bluetooth") }
                Unit
            }
        }

        // --- Source BLE (allowlist de services "UART série") ---
        if (enabled("ble") && btAdapter != null) {
            jobs += async {
                runCatching {
                    BleScanner(context, btAdapter).scan(options.timeoutMs, collect)
                }.onFailure { failed.add("ble") }
                Unit
            }
        }

        jobs.awaitAll()

        Pair(merge(buffer.toList()), failed.distinct())
    }

    /** Fusion + dédoublonnage par id stable, en conservant le meilleur adapter. */
    private fun merge(incoming: List<DiscoveredPrinter>): List<DiscoveredPrinter> {
        val byId = LinkedHashMap<String, DiscoveredPrinter>()
        for (p in incoming) {
            val existing = byId[p.id]
            if (existing == null) {
                byId[p.id] = p
                continue
            }
            val mergedSources = (existing.discoveredBy + p.discoveredBy).toMutableSet()
            val winner = if (AdapterPriority.score(p) > AdapterPriority.score(existing)) p else existing
            winner.discoveredBy.clear()
            winner.discoveredBy.addAll(mergedSources)
            winner.lastSeenAt = maxOf(existing.lastSeenAt, p.lastSeenAt)
            winner.isConnected = existing.isConnected || p.isConnected
            byId[p.id] = winner
        }
        return byId.values.sortedWith(
            compareByDescending<DiscoveredPrinter> { AdapterPriority.score(it) }.thenBy { it.name },
        )
    }
}

File: ios/Plugin/Discovery/DiscoveryManager.swift
import Foundation

/// Orchestre la découverte agrégée iOS (miroir de DiscoveryManager.kt).
///
/// Sources iOS :
///   - SDK Epson / Star / Brother / Zebra (via adapters, si liés)
///   - Bonjour/mDNS réseau (BonjourScanner)
///   - BLE (optionnel, allowlist)
///
/// PAS de Bluetooth Classic générique sur iOS (voir README "Limites iOS").
final class DiscoveryManager {

    struct Options {
        let sources: Set<String>?
        let timeoutMs: Int
        let networkCidr: String?
        let tcpPorts: [Int]
    }

    private let adapters: [PrinterAdapter]
    init(adapters: [PrinterAdapter]) { self.adapters = adapters }

    func discover(_ options: Options, emitPartial: @escaping (DiscoveredPrinter) -> Void) async -> (printers: [DiscoveredPrinter], failed: [String]) {
        let buffer = SyncBuffer()
        var failed: [String] = []

        func enabled(_ src: String) -> Bool { options.sources == nil || options.sources!.contains(src) }

        let collect: (DiscoveredPrinter) -> Void = { p in
            buffer.add(p)
            emitPartial(p)
        }

        await withTaskGroup(of: Void.self) { group in
            // SDK fabricants
            for adapter in adapters {
                let src: String?
                switch adapter.id {
                case .epson: src = "epson"
                case .star: src = "star"
                case .brother: src = "brother"
                case .zebra: src = "zebra"
                default: src = nil
                }
                if let src = src, enabled(src), adapter.isAvailable() {
                    group.addTask { await adapter.discover(timeoutMs: options.timeoutMs, onFound: collect) }
                }
            }
            // Bonjour réseau
            if enabled("tcp") {
                group.addTask { await BonjourScanner().scan(timeoutMs: options.timeoutMs, onFound: collect) }
            }
            // BLE (optionnel)
            // if enabled("ble") { group.addTask { await BleScanner().scan(...) } }
        }

        return (merge(buffer.snapshot()), failed)
    }

    private func merge(_ incoming: [DiscoveredPrinter]) -> [DiscoveredPrinter] {
        var byId: [String: DiscoveredPrinter] = [:]
        for p in incoming {
            if let existing = byId[p.id] {
                var winner = AdapterPriority.score(p) > AdapterPriority.score(existing) ? p : existing
                winner.discoveredBy = existing.discoveredBy.union(p.discoveredBy)
                winner.lastSeenAt = max(existing.lastSeenAt, p.lastSeenAt)
                winner.isConnected = existing.isConnected || p.isConnected
                byId[p.id] = winner
            } else {
                byId[p.id] = p
            }
        }
        return byId.values.sorted {
            let sa = AdapterPriority.score($0), sb = AdapterPriority.score($1)
            return sa != sb ? sa > sb : $0.name < $1.name
        }
    }
}

/// Petit buffer thread-safe pour collecter les résultats concurrents.
final class SyncBuffer {
    private var items: [DiscoveredPrinter] = []
    private let lock = NSLock()
    func add(_ p: DiscoveredPrinter) { lock.lock(); items.append(p); lock.unlock() }
    func snapshot() -> [DiscoveredPrinter] { lock.lock(); defer { lock.unlock() }; return items }
}
